internal/services/gateway/action/project.go

```diff
@@ -584,3 +591,40 @@ func (h *ActionHandler) getRemoteRepoAccessData(ctx context.Context, linkedAccou
 
 	return user, rs, la, nil
 }
+
+func (h *ActionHandler) GetRepoInfo(ctx context.Context, repoPath string, remoteSourceRef string) (*gitsource.RepoInfo, error) {
+	curUserID := common.CurrentUserID(ctx)
+	gitSource, _, _, err := h.GetGitSourceByRemoteSourceRef(ctx, remoteSourceRef, curUserID)
+	if err != nil {
+		return nil, util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to get remote source %q", remoteSourceRef))
+	}
+
+	repo, err := gitSource.GetRepoInfo(repoPath)
+	if err != nil {
+		return nil, util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to get repository info from gitsource %q", repoPath))
+	}
+
+	return repo, nil
+}
+
+func (h *ActionHandler) RefreshRemoteRepositoryInfo(ctx context.Context, projectRef string) (*csapitypes.Project, error) {
+	project, err := h.GetProject(ctx, projectRef)
+	if err != nil {
+		return nil, util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to get project %q", projectRef))
+	}
+
+	repoInfo, err := h.GetRepoInfo(ctx, project.RepositoryPath, project.RemoteSourceID)
+	if err != nil {
+		return nil, util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to get repo info %q", project.RepositoryPath))
+	}
+
+	areq := &UpdateProjectRequest{
+		DefaultBranch: &repoInfo.DefaultBranch,
+	}
+	project, err = h.UpdateProject(ctx, projectRef, areq)
```

Comment:
It's not really cleaner to call an user action from another action. Just create a not exported `updateProject` function removing the permission checking logic from it and call it from here and from the `UpdateProject` method. Put the permission logic in the main actions methods.